sys/include/net/asymcute.h

```diff
@@ -0,0 +1,563 @@
+    uint8_t rxbuf[ASYMCUTE_BUFSIZE];    /**< connection specific receive buf */
+    char cli_id[ASYMCUTE_ID_MAXLEN + 1];    /**< buffer to store client ID */
+};
+
+/**
+ * @brief   Data-structure for holding topics and their registration status
+ */
+struct asymcute_topic {
+    asymcute_con_t *con;        /**< connection used for registration */
+    char name[ASYMCUTE_TOPIC_MAXLEN + 1];   /**< topic string (ACSII only) */
+    uint8_t flags;              /**< normal, short, or pre-defined */
+    uint16_t id;                /**< topic id */
+};
+
+/**
+ * @brief   Data-structure holding the state of subscriptions
+ */
+struct asymcute_sub {
+    asymcute_sub_t *next;       /**< the subscriptions list entry */
+    asymcute_topic_t *topic;    /**< topic we subscribe to */
+    asymcute_sub_cb_t cb;       /**< called on incoming data */
+    void *arg;                  /**< user supplied callback argument */
+};
+
+/**
+ * @brief   Data structure for defining a last will
+ */
+struct asymcute_will {
+    const char *topic;          /**< last will topic */
+    void *msg;                  /**< last will message content */
+    size_t msg_len;             /**< length of last will message content */
+};
+
+/**
+ * @brief   Check if a given request context is currently used
+ *
+ * @param[in] req       request context to check
+ *
+ * @return  true if context is currently used
+ * @return  false if context is not used
+ */
+static inline bool asymcute_req_in_use(const asymcute_req_t *req)
+{
+    assert(req);
+    return (req->con != NULL);
+}
+
+/**
+ * @brief   Check if a given subscription is currently active
+ *
+ * @param[in] sub       subscription context to check
+ *
+ * @return  true if subscription is active
+ * @return  false if subscription is not active
+ */
+static inline bool asymcute_sub_active(const asymcute_sub_t *sub)
+{
+    assert(sub);
+    return (sub->topic != NULL);
+}
+
+/**
+ * @brief   Reset the given topic
+ *
+ * @note    Make sure that the given topic is not used by any subscription or
+ *          last will when calling this function
+ *
+ * @param[out] topic    topic to reset
+ */
+static inline void asymcute_topic_reset(asymcute_topic_t *topic)
+{
+    assert(topic);
+    memset(topic, 0, sizeof(asymcute_topic_t));
+}
+
+/**
+ * @brief   Check if a given topic is currently registered with a gateway
+ *
+ * @param[in] topic     topic to check
+ *
+ * @return  true if topic is registered
+ * @return  false if topic is not registered
+ */
+static inline bool asymcute_topic_is_reg(const asymcute_topic_t *topic)
+{
+    assert(topic);
+    return (topic->con != NULL);
+}
+
+/**
+ * @brief   Check if a given topic is initialized
+ *
+ * @param[in] topic     topic to check
+ *
+ * @return  true if topic is initialized
+ * @return  false if topic is not initialized
+ */
+static inline bool asymcute_topic_is_init(const asymcute_topic_t *topic)
+{
+    assert(topic);
+    return (strlen(topic->name) > 0);
```

Comment:
You can safe a few byte of ROM (on `nrf52dk` and `samr21-xpro` I've got 16 bytes, 22 bytes on `iotlab-m3`) and the `<string.h>` include by doing `topic->name[0] != '\0'` here.